### PR TITLE
volumegroup: add volumeID's to createGroup

### PR DESCRIPTION
**Pull request by Madhu-1**
add an optional volumeID's to the CreateVolumeGroupRequest where some storage providers might not support creating a empty group and expect some volumes to exist to create a group. and also helps to pre-identify the group details for which volume already exists where a volume can be part of only one group and cannot be part of multiple groups.

### CODE REVIEW

**Comment by nixpanic** [comment]
hmm, I think it would be  better to remove volumgroup support and have CSI-drivers handle internally how/if they need volumegroups.

kubernetes-csi-addons will just pass listst of volumes to CSI-Addons procedures, just like VolumeGroupSnapshot does in its related CSI procedures.

**Comment by Madhu-1** [comment]
> hmm, I think it would be better to remove volumgroup support and have CSI-drivers handle internally how/if they need volumegroups.
> 
> kubernetes-csi-addons will just pass listst of volumes to CSI-Addons procedures, just like VolumeGroupSnapshot does in its related CSI procedures.

@nixpanic This is a complete change in plan for the implementation the plan was to reuse the volumegroup concept under the volumegroupreplication concept where the user will not see the volumegroup but we will make use of the volumegroup concept. any specific reason for this suggestion.

**Comment by nixpanic** [comment]
> > hmm, I think it would be better to remove volumgroup support and have CSI-drivers handle internally how/if they need volumegroups.
> > kubernetes-csi-addons will just pass listst of volumes to CSI-Addons procedures, just like VolumeGroupSnapshot does in its related CSI procedures.
> 
> @nixpanic This is a complete change in plan for the implementation the plan was to reuse the volumegroup concept under the volumegroupreplication concept where the user will not see the volumegroup but we will make use of the volumegroup concept. any specific reason for this suggestion.

CSI-drivers can use a volumegroup structure internally if they need one. I don't think it should be part of an API for consumers (like a CSI-Addons API).

**Comment by Madhu-1** [comment]
> CSI-drivers can use a volumegroup structure internally if they need one. I don't think it should be part of an API for consumers (like a CSI-Addons API).

what is the use of it? the earlier discussion was to separate the replication and grouping into two different parts but now the suggestion is to make it a single API for both operations, am I assuming it is correct or you are suggesting something else?

**Review by Rakshith-R — APPROVED** [review]

**Comment by Madhu-1** [comment]
@nixpanic PTAL based on the discussion as we agreed to keep the current flow

File: volumegroup/README.md [hunk 1]
```diff
@@ -47,7 +47,7 @@ If a volume group corresponding to the specified volume group name already exist
 parameters in the CreateVolumeGroupRequest, the Plugin MUST reply 0 OK with the corresponding CreateVolumeGroupResponse.
 CSI Plugins MAY create the following types of volume groups:
 
-Create a new empty volume group. Note that N volumes with some backend label Y could be considered to be in "group Y"
+Create a new empty volume group or a group with specific volumes. Note that N volumes with some backend label Y could be considered to be in "group Y"
 which might not be a physical group on the storage backend. In this case, an empty group can still be created by the CO
 to hold volumes. After the empty group is created, create a new volume, specifying the group name in the volume.
```
Comment (nixpanic) [inline volumegroup/README.md:52]: This isn't an option, maybe remove it?

> create a new volume, specifying the group name in the volume

Or explain that `ModifyVolumeGroupMembership` needs to be called to add the volume afterwards.

File: volumegroup/README.md [hunk 2]
```diff
@@ -78,6 +75,10 @@ message CreateVolumeGroupRequest {
   // This field is OPTIONAL. Refer to the `Secrets Requirements`
   // section on how to use this field.
   map<string, string> secrets = 3 [(csi.v1.csi_secret) = true];
+
+  // Specify volume_ids that will be added to the volume group.
+  // This field is OPTIONAL.
+  repeated string volume_ids = 4;
```
Comment (nixpanic) [inline volumegroup/README.md:81]: there should be an error code and description that explains an incompatible volume_ids list is passed

**Review by nixpanic — CHANGES_REQUESTED** [review]

File: volumegroup/README.md [hunk 3]
```diff
@@ -47,9 +47,10 @@ If a volume group corresponding to the specified volume group name already exist
 parameters in the CreateVolumeGroupRequest, the Plugin MUST reply 0 OK with the corresponding CreateVolumeGroupResponse.
 CSI Plugins MAY create the following types of volume groups:
 
-Create a new empty volume group. Note that N volumes with some backend label Y could be considered to be in "group Y"
+Create a new empty volume group or a group with specific volumes. Note that N volumes with some backend label Y could be considered to be in "group Y"
 which might not be a physical group on the storage backend. In this case, an empty group can still be created by the CO
-to hold volumes. After the empty group is created, create a new volume, specifying the group name in the volume.
+to hold volumes. After the empty group is created, create a new volume,
+specifying the group name in the volume. CO may call ModifyVolumeGroupMembership to add new volumes to the group.
```
Comment (nixpanic) [inline volumegroup/README.md:53]: This is not an option, a CSI Volume does not know about VolumeGroups:

> specifying the group name in the volume

Please drop that part.

File: volumegroup/README.md [hunk 4]
```diff
@@ -47,9 +47,10 @@ If a volume group corresponding to the specified volume group name already exist
...
 
 At restore time, create a single volume from individual snapshot and then join an existing group. Create an empty group.
 Create a volume from snapshot, specifying the group name in the volume.
```
Comment (nixpanic) [inline volumegroup/README.md:56]: Same bit is mentioned here too.

File: volumegroup/README.md (hunk 2 above)
Comment (nixpanic) [reply volumegroup/README.md:82]: I was expecting an `INVALID_ARGUMENT` error or something new... If the volume can only be part of a single group, returning `ALREADY_EXISTS` and asking to use a different `name` is not a solution.

Comment (Madhu-1) [reply volumegroup/README.md:81]: i referred to https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolumegroupsnapshot-errors for it, i can add invalid if you prefer but single volume is a limitation from the SP, let me know if you still suggest to have invalid error

Comment (nixpanic) [reply volumegroup/README.md:81]: I prefer `FAILED_PRECONDITION` for volumes that can not be grouped together, for whatever reason (already part of a group, different features, ...). The `ALREADY_EXISTS` can then be a clear error that a volumegroup with the same name already exists (and the caller can then check if the volumegroup actually has the expected volumes).

Comment (Madhu-1) [reply volumegroup/README.md:81]: @nixpanic PTAL

**Review by nixpanic — APPROVED** [review]
Thanks!

**Review by Rakshith-R — APPROVED** [review]

**Pull request merged**